plugins/c-ares-dns-client.c

```diff
@@ -0,0 +1,757 @@
+	}
+
+	memcpy(addr_buffer, hostent->h_addr, hostent->h_length);
+
+	if (inet_ntop(hostent->h_addrtype, addr_buffer, ip_addr_string,
+		      sizeof(ip_addr_string)) == NULL) {
+		DBG("could not convert address to string: %s",
+				strerror(errno));
+		request->cb(request->data, OFONO_DNS_CLIENT_ERROR_NO_DATA,
+			    NULL);
+		return;
+	}
+
+	DBG("ares request for '%s' succeeded with %d timeouts: %s",
+			request->hostname, timeouts, ip_addr_string);
+	request->cb(request->data, status_from_ares_status(ares_status),
+		    ip_addr);
+}
+
+/* Cancel all in-progress asynchronous name resolution requests. */
+static void cancel_all_ares_requests()
+{
+	GList *node;
+	struct ares_request *request;
+
+	DBG("");
+
+	while ((node = g_list_first(pending_requests)) != NULL) {
+		request = node->data;
+		pending_requests = g_list_delete_link(pending_requests, node);
+		request->running = FALSE;  /* don't trip assertion */
+		destroy_ares_request(request);
+	}
+}
+
+static void set_request_servers(const char **servers,
+						struct ares_request *request)
+{
+	struct ares_addr_node *current;
+	struct ares_addr_node **prev = &request->servers;
+	const char **dns;
+	int rc;
+
+	for (dns = servers; *dns != NULL; ++dns) {
+		current = g_malloc0(sizeof(*current));
+		rc = inet_pton(AF_INET, *dns, &current->addr.addr4);
+		if (rc == 1) {
+			current->family = AF_INET;
+		} else {
+			rc = inet_pton(AF_INET6, *dns, &current->addr.addr6);
+			if (rc != 1) {
+				ofono_error("%s: wrong dns address %s",
+								__func__, *dns);
+				g_free(current);
+				continue;
+			}
+			current->family = AF_INET6;
+		}
+
+		*prev = current;
+		prev = &current->next;
+	}
+}
+
+/* Initiate an asynchronous name resolution request. */
+static ofono_dns_client_request_t
+ofono_dns_client_submit_request(const char *hostname,
+				const char *interface,
+				const char **servers,
+				int timeout_ms,
+				ofono_dns_client_callback_t cb,
+				void *data)
+{
+	int ares_status;
+	struct ares_request *request;
+	struct ares_options options;
+	int optmask;
+	const gboolean destroy_old_source = TRUE;
+
+	DBG("");
+
+	if (timeout_ms < 0) {
+		DBG("invalid timeout value of %d ms", timeout_ms);
+		return NULL;
+	}
+
+	request = g_malloc0(sizeof(struct ares_request));
+	request->running = TRUE;
+
+	request->ares_watches = g_hash_table_new_full(g_int_hash, g_int_equal,
+						      NULL, destroy_ares_watch);
+	if (request->ares_watches == NULL) {
+		DBG("could not create ares_watches table");
+		g_free(request);
+		return NULL;
+	}
+
+	/*
+	 * Init a c-ares channel for this request. We set an option asking
+	 * c-ares to notify us via callback about which sockets it wants to
+	 * monitor for readability and writability. This allows us to
+	 * integrate c-ares activity into our glib main event loop.
+	 */
+	memset(&options, 0, sizeof(options));
+	options.sock_state_cb = ares_socket_state_cb;
+	options.sock_state_cb_data = request;
+	optmask = ARES_OPT_SOCK_STATE_CB;
+	if (timeout_ms > 0) {
```

Comment:
We should discuss whether or not to keep the original formatting or apply the ofono coding style...